### PR TITLE
Fix bug - fail to apply coupon if it is only applicable to specific shipping method

### PR DESCRIPTION
**Pull request by dumega**
asana task https://app.asana.com/0/941920570700290/1118193263476543/f

### CODE REVIEW

File: Model/Api/DiscountCodeValidation.php [hunk 1]
```diff
@@ -708,8 +758,8 @@ private function convertToBoltDiscountType($type)
     private function setCouponCode($quote, $couponCode)
     {
         $quote->getShippingAddress()->setCollectShippingRates(true);
-        $quote->setCouponCode($couponCode);
-        $this->cartHelper->saveQuote($quote->collectTotals());
+        $quote->setCouponCode($couponCode)->collectTotals()->save();
+        $this->cartHelper->saveQuote($quote);
```
Comment (YevhenShBolt) [inline Model/Api/DiscountCodeValidation.php]: I think we should not save twice the same object!

Comment (YevhenShBolt) [reply Model/Api/DiscountCodeValidation.php]: What is the reason to add here
`$quote->setCouponCode($couponCode)->collectTotals()->save();`
additional save ?

**Comment by danac-gs** [comment]
@dumega What I meant by testing was not only unit tests but running manual production readiness alike tests on a store with both simple coupons / shipping options and more complicated rules specific to the task and various shipping carriers. My concern is we don't break something that used to work. But I expect you already did that.

**Comment by dumega** [comment]
@danac-gs  @YevhenShBolt @matt-thomason I have run some more tests for this PR today, no issue found for now.

and @YevhenShBolt , for your question,
1. I remove the duplicate save `$this->cartHelper->saveQuote($quote);` and only keep `$quote->setCouponCode($couponCode)->collectTotals()->save();`

2. Then I found that `$this->cartHelper->saveQuote($quote);` does not detect the conditions for coupons if the conditions are applied to shipping method only.

Then I will add unit tests

**Pull request merged**